ccmlib/common.py

```diff
@@ -986,40 +1000,36 @@ def _update_java_version(current_java_version, current_java_home_version,
-                .format(info_message, current_java_version, cassandra_version, supported_versions))
-
-            # nothing to do, java is on the path and JAVA_HOME points to the distribution of the same version
+            # nothing to change in this case
             return env
 
-        elif current_java_home_version:
-            info('{}: Using the current Java version {} from JAVA_HOME for the current invocation of Cassandra {}.'
-                 .format(info_message, current_java_home_version, cassandra_version))
-            if current_java_home_version not in supported_versions:
-                warning('{}: The current Java version {} from JAVA_HOME is not supported by Cassandra {} (supported versions: {}).'
-                        .format(info_message, current_java_home_version, cassandra_version, supported_versions))
-
-            # nothing more to be done, just add JAVA_HOME/bin to the PATH
+        elif current_java_home_version and current_java_home_version in supported_versions:
+            info('{}: Using the current Java {} available from JAVA_HOME for the current invocation of Cassandra {}.'
+                 .format(info_message, current_java_version, cassandra_version))
+            # just need to add JAVA_HOME/bin to the PATH
             return update_path_in_env(env, env['JAVA_HOME'] + '/bin')
 
-        # Fallback - use the highest supported Java version available in the current env
-        elif len(supported_available_versions) > 0:
-            jvm_version = sorted(supported_available_versions, reverse=True)[0]
-            info('{}: Using the highest supported Java version {} available at {} for the current invocation of Cassandra {}.'
-                 .format(info_message, jvm_version, env[available_versions[jvm_version]], cassandra_version))
-            env['JAVA_HOME'] = env[available_versions[jvm_version]]
-            return update_path_in_env(env, env['JAVA_HOME'] + '/bin')
+        elif current_java_home_version and current_java_home_version not in supported_versions:
+            warning('{}: The current Java {} is not supported by Cassandra {} (supported versions: {}).'
+                    .format(info_message, current_java_version, cassandra_version, supported_versions))
+        else:
+            warning('{}: JAVA_HOME is not defined; CCM will try to find a suitable Java distribution among JAVAx_HOME env variables.')
+
+        if len(supported_available_versions) > 0:
+            info('{}: CCM has found {} Java distributions, the required Java version for Cassandra {} is {}.'.format(
+                info_message, str(available_versions), cassandra_version, supported_versions))
+            ref_version = current_java_home_version if current_java_home_version else max(supported_available_versions)
+            # sort by the absolute difference between the current java version and the supported version
+            # to use the closest version if the current version is not supported
```

Comment:
nit: the default jdk for any C* version is always its lowest jdk
🤷

(the rationale here to be different is that we're assuming the closest will involve less jdk-switching over possible upgrade steps (the caller is doing, e.g. dtest-upgrade-large) and therefore enable more such steps.  i'm not entirely sure this is the best approach, it might be more intuitive that C* versions always fall back to preferring their default jdk…)